sympy/polys/euclidtools.py

```diff
@@ -1248,6 +1247,386 @@ def dup_zz_heu_gcd(f, g, K):
+    variable_range = range(num_variables)
+
+    for monomial, coefficient in p1.items():
+        symbol_indices = set(compress(variable_range, monomial))
+        monomial_with_symbols = [0] * num_variables
+        monomial_without_symbols = [0] * num_variables
+
+        for i in symbol_indices & symbols:
+            monomial_with_symbols[i] = monomial[i]
+
+        for i in symbol_indices & non_symbol_indices:
+            monomial_without_symbols[i] = monomial[i]
+
+        p2[tuple(monomial_with_symbols)][tuple(monomial_without_symbols)] = coefficient
+
+    return p2
+
+
+def get_sparse_coefficients(p1, symbols):
+    """
+    Compute the sparse coefficients of a polynomial with respect to a set of symbols.
+    """
+    p2 = compute_sparse_coefficients(p1, symbols)
+    return [p1.ring(pi) for pi in p2.values()]
+
+
+def get_sparse_coefficient(p1, symbol, degree):
+    """
+    Returns the coefficient of a specific term in a polynomial p1, given the symbol and degree of the term.
+    """
+    p2 = compute_sparse_coefficients(p1, {symbol})
+    monomial = [0] * len(p1.ring.gens)
+    monomial[symbol] = degree
+    monomial = tuple(monomial)
+    return p1.ring(p2[monomial])
+
+
+def compute_gcd_sparse(p):
+    """
+    Computes the greatest common divisor (GCD) of a list of polynomials p with symbolic coefficients.
+    """
+    ring = p[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in p):
+        return compute_gcd_terms(p, ring, domain)
+
+    p, monomial_gcd = compute_gcd_monomial(p)
+
+    p, common_symbols = compute_gcd_coefficients(p)
+
+    gcd = p[0]
+    for pi in p[1:]:
+        gcd = compute_gcd_prs_sparse(gcd, pi)
+
+    if monomial_gcd is not None:
+        gcd = gcd * monomial_gcd
+
+    return gcd
+
+
+def compute_gcd_terms(p, ring, domain):
+    """
+    Computes the greatest common divisor (GCD) of a list of polynomials p in a given ring and domain.
+    """
+    terms = chain.from_iterable(pi.terms() for pi in p)
+    monomials = set()
+    coefficients = set()
+
+    for monomial, coefficient in terms:
+        monomials.add(monomial)
+        coefficients.add(coefficient)
+
+    monomial_gcd = compute_monomial_gcd(monomials)
+    coefficient_gcd = compute_ground_gcd(coefficients, domain)
+    term_gcd = ring({monomial_gcd: coefficient_gcd})
+
+    return term_gcd
+
+
+def compute_ground_gcd(coefficients, domain):
+    """
+    Computes the greatest common divisor (GCD) of multiple polynomials represented by their terms.
+    """
+    coefficients = list(coefficients)
+    gcd = domain.gcd
+    d = coefficients[0]
+
+    for coefficient in coefficients[1:]:
+        d = gcd(d, coefficient)
+
+        if d == domain.one:
+            break
+
+    return d
+
+
+def compute_monomial_gcd(monomials):
+    """
+    Computes the greatest common divisor (GCD) of the exponents for each variable in the monomials.
+    """
+    monomial_gcd = tuple(map(min, zip(*monomials)))
+    return monomial_gcd
```

Comment:
If there are many terms and many monomials this would loop over all of them. We could check beforehand if the constant monomial is present because then we don't need to look at any of the others to find the gcd.

It would be best to check this when we still have a dict representation because the check `(0, 0, 0, ...) in p` is `O(1)`.